### DISCUSSION TITLE
Correct way to implement bi-level optimization

### DISCUSSION THREAD

**Discussion by MichaelMMeskhi** [jax-ml/jax#13551] [Q&A] [answered]
Hello, I am having trouble implementing a bi-level optimization correctly. Assume I fit a model on some data, I would like to update another model w.r.t. the loss of the base model. The outer model outputs some data that the base model uses hence there is a bi-level dependence. 

```python3
def train(outer_state, inner_state, rnd_vector, input, target):
    def outer_loss_fn(outer_params):
        def inner_loss_fn(params):
            logits = InnerModel().apply({'params': params}, input)
            loss = aug_cross_entropy_loss(logits=logits, labels=targets, z)
            return loss
        
        z = OuterModel().apply({'params': outer_params}, rnd_vector)

        for _ in range(5): # k-steps of gradient updates
            inner_grad_fn = jax.value_and_grad(inner_loss_fn, has_aux=False)
            loss, inner_grads = inner_grad_fn(inner_state.params)
            updated_inner_state = inner_state.apply_gradients(grads= inner_grads)

        return loss, updated_inner_state

    outer_grad_fn = jax.value_and_grad(outer_loss_fn, has_aux=True)
    (loss, updated_inner_state), outer_grads = outer_grad_fn(outer_state.params)
    updated_outer_state = outer_state.apply_gradients(grads=outer_grads)
    return updated_outer_state, updated_inner_state, loss
```

The issue shows up at line 

```python3
(loss, updated_inner_state), outer_grads = outer_grad_fn(outer_state.params)
```

After 1 iteration, gradients end up as `nans`.

**Comment by pharringtonp19** [comment, marked as answer]
@MichaelMMeskhi from a quick look at this, it looks like `outer_params` is not used within `outer_loss_fn`.

**Reply by MichaelMMeskhi** [reply]
My apologies, I had a typo. It is used to generate `z`.

**Reply by pharringtonp19** [reply]
@MichaelMMeskhi Any chance you could share this as a colab notebook with everything defined? I think it would be easier to work through that way. It looks like `base_state` is also not defined.

**Reply by MichaelMMeskhi** [reply]
Unfortunately, I don't think I have the permission to share that much. Regarding `base_state`, again that's a typo let me fix it. Let me see if there is more information I can provide.

**Reply by jakevdp** [reply]
You might go through the exercise of creating a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example). Replace proprietary things with random data and minimize until you have a reproduction that others can run. Otherwise I doubt you'll get any useful answers to your question: it's just difficult to guess from a partial code snippet where an error like this might be coming from

Here's one issue that might be relevant: https://jax.readthedocs.io/en/latest/faq.html#gradients-contain-nan-where-using-where

**Reply by MichaelMMeskhi** [reply]
I hope [this can be of use](https://colab.research.google.com/drive/1gG69i5lJPrrEI8x4MOVy0k02TEWHSRVM?usp=sharing)

**Reply by jakevdp** [reply]
Thanks, but the function contains a syntax error and the notebook does not show how it is called (I don't know what `rnd_vector` and `target` are supposed to be)

**Reply by MichaelMMeskhi** [reply]
Sorry there was a missing code block(s). It is fully functional now.

**Reply by jakevdp** [reply]
Thanks - it seems to run without any problems. Do you still have a question about this code?

**Reply by MichaelMMeskhi** [reply]
Yes, my question is, the way I have it implemented, is the outer model's parameters being correctly updated wrt to the performance of the inner model?

**Reply by jakevdp** [reply]
I guess so? Do you have any reason to believe it isn't?
